cmd/internal/kube/daemonset/apidump_process.go

```diff
@@ -0,0 +1,104 @@
+	"github.com/postmanlabs/postman-insights-agent/printer"
+	"github.com/postmanlabs/postman-insights-agent/rest"
+	"github.com/postmanlabs/postman-insights-agent/telemetry"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func (d *Daemonset) StartApiDumpProcess(podUID types.UID) error {
+	podArgs, err := d.getPodArgsFromMap(podUID)
+	if err != nil {
+		return err
+	}
+
+	err = podArgs.changePodTrafficMonitorState(TrafficMonitoringStarted, PodDetected, PodInitialized)
+	if err != nil {
+		return errors.Wrapf(err, "failed to change pod state, pod name: %s, from: %d to: %d",
+			podArgs.PodName, podArgs.PodTrafficMonitorState, TrafficMonitoringStopped)
+	}
+
+	go func() (funcErr error) {
+		// defer function handle the error (if any) in the apidump process and change the pod state accordingly
+		defer func() {
+			nextState := TrafficMonitoringEnded
+
+			if err := recover(); err != nil {
+				printer.Errorf("Panic occurred in apidump process for pod %s, err: %v\n%v\n",
+					podArgs.PodName, err, string(debug.Stack()))
+				nextState = TrafficMonitoringFailed
+			} else if funcErr != nil {
+				printer.Errorf("Error occurred in apidump process for pod %s, err: %v", podArgs.PodName, funcErr)
+				nextState = TrafficMonitoringFailed
+			} else {
+				printer.Infof("Apidump process ended for pod %s", podArgs.PodName)
+			}
+
+			err = podArgs.changePodTrafficMonitorState(nextState, TrafficMonitoringStarted)
+			if err != nil {
+				printer.Errorf("Failed to change pod state, pod name: %s, from: %d to: %d, error: %v",
+					podArgs.PodName, podArgs.PodTrafficMonitorState, nextState, err)
+				return
+			}
+
+			// It is possible that the apidump process is already stopped and the stopChannel is of no use
+			// This is just a safety check
+			err := d.StopApiDumpProcess(podUID, err)
+			if err != nil {
+				printer.Errorf("Failed to stop api dump process, pod name: %s, error: %v", podArgs.PodName, err)
+			}
+		}()
+
+		networkNamespace, err := d.CRIClient.GetNetworkNamespace(podArgs.ContainerUUID)
+		if err != nil {
+			funcErr = errors.Errorf("Failed to get network namespace for pod/containerUUID: %s/%s, err: %v",
+				podArgs.PodName, podArgs.ContainerUUID, err)
+			return
+		}
+
+		apidumpArgs := apidump.Args{
+			ClientID:  telemetry.GetClientID(),
+			Domain:    rest.Domain,
+			ServiceID: podArgs.InsightsProjectID,
+			ReproMode: d.InsightsReproModeEnabled,
+			DaemonsetArgs: optionals.Some(apidump.DaemonsetArgs{
+				TargetNetworkNamespaceOpt: networkNamespace,
+				StopChan:                  podArgs.StopChan,
+				APIKey:                    podArgs.PodCreds.InsightsAPIKey,
+				Environment:               podArgs.PodCreds.InsightsEnvironment,
+			}),
+		}
+
+		if err := apidump.Run(apidumpArgs); err != nil {
+			funcErr = errors.Wrapf(err, "Failed to run apidump process for pod %s", podArgs.PodName)
+		}
+		return
+	}()
+
+	return nil
+}
+
+func (d *Daemonset) StopApiDumpProcess(podUID types.UID, stopErr error) error {
```

Comment:
This is non-blocking, which is fine.

But that means that during daemonset shutdown we exit the main process before the apidump processes have had a chance to shut down gracefully.  Is there something we can do wait for them (for some bounded time period) before exiting, in StopAllApidumpProcesses?